WD-Firmware/comrv/comrv_entry.S

```diff
@@ -0,0 +1,180 @@
+    /* change sp for temporary storage */
+    m_ADDI  sp, sp, -D_COMRV_STACK_SIZE
+    /* temporary save a0 in the next free stack frame */
+    m_STORE a0, D_COMRV_A0_REG_STACK_OFFSET_TMP(sp)
+    /* prepare address output arg for comrvGerFunctionAddressFromToken - overlay loaded
+       to comrv table entry */
+    m_ADDI  a0, t3, D_COMRV_GROUP_SIZE_STACK_OFFSET
+    /* search if the token (t5) is already loaded */
+    jal     comrvGerFunctionAddressFromToken
+    /* save the loaded function address in t5 (instead of the token) */
+    mv      t5, a0
+    /* restore a0 - may contain an actual callee function argument */
+    m_LOAD  a0, D_COMRV_A0_REG_STACK_OFFSET_TMP(sp)
+    /* adjust back sp - memory not needed anymore */
+    m_ADDI  sp, sp, D_COMRV_STACK_SIZE
+comrv_invoke_callee:
+    /* call the function */
+    jalr    t5
+    /* load the return address */
+    m_LOAD  ra, D_COMRV_CALLER_RA_STACK_OFFSET(t3)
+    /* we need to load the caller token to detrmine if the RA is an offset
+       or an actual return address; first we need to unlink the top of the
+       com-rv stack frame, return it back to the stack frame pool and get
+       read the caller token (in previous frame it will bethe  callee token) */
+    /* load the offset to previous stack frame */
+    lh      t5, D_COMRV_PREV_STACK_FRAME(t3)
+    /* unlink the top of the stack - t3 hold the unlinked stack frame address
+       t5 will hold the address of the top of the stack */
+    add     t5, t5, t3
+    /* disable interrupts */
+    m_COMRV_DISABLE_INTS
+    /* link the freed stack frame back to the stack frames pool;
+       t2 will hold the offset to the previous free stack frame */
+    sub     t4, t4, t3
+    /* save the offset to the previous free stack frame */
+    sh      t4, D_COMRV_PREV_STACK_FRAME(t3)
+    /* link the free stack frame back to the pool */
+    mv      t4, t3
+    /* enable interrupts */
+    m_COMRV_ENABLE_INTS
+    /* save the top of the stack to the stack register */
+    mv      t3, t5
+    /* read caller token (will be in the callee field now as we look at
+       the previous stack frame) */
+    m_LOAD  t5, D_COMRV_CALLEE_TOKEN_STACK_OFFSET(t3)
+    /* t2 will hold the token type - address/token */
+    andi    t2, t5, 1
+    /* t2 is 0 means return to a non-overlay function so we can skip
+       token search and load */
+    beqz    t2, comrv_exit_ret_to_caller
+    /* change sp for temporary storage */
+    m_ADDI  sp, sp, -D_COMRV_STACK_SIZE
+    /* temporary save a0 as it is being used */
+    m_STORE a0, D_COMRV_A0_REG_STACK_OFFSET_TMP(sp)
+    /* temporary save ra as it is being used */
+    m_STORE ra, D_COMRV_RA_REG_STACK_OFFSET_TMP(sp)
+    /* the arg passed to comrvGerFunctionAddressFromToken isn't used now so
+       we just provide it a temp location on the stack */
+    m_ADDI  a0, sp, D_COMRV_ARG_STACK_OFFSET_TMP
+    /* search if the token (t5) is already loaded */
+    jal     comrvGerFunctionAddressFromToken
```

Comment:
on comment: fix ger to get
on code: think on return the "real address" and not doing manipulation for offset. do it inside the function